File: guests/haskell/align.hs
-- The align scene, Haskell port — guests/rust/align.rs,
-- tools/scenes/align.steps.

import qualified Data.ByteString as BS
import KayaApp
import KayaWire (Value (..))

-- A 2x64 PNG: the tall no-baseline child.
tallPng :: BS.ByteString
tallPng =
  BS.pack
    [ 137, 80, 78, 71, 13, 10, 26, 10, 0, 0, 0, 13,
      73, 72, 68, 82, 0, 0, 0, 2, 0, 0, 0, 64,
      8, 2, 0, 0, 0, 191, 68, 49, 20, 0, 0, 0,
      18, 73, 68, 65, 84, 120, 156, 99, 8, 8, 138, 2,
      34, 134, 81, 106, 104, 82, 0, 67, 50, 126, 1, 49,
      1, 65, 124, 0, 0, 0, 0, 73, 69, 78, 68, 174,
      66, 96, 130
    ]

main :: IO ()
main = kayaMain $ \app -> do
  buildTx app $ do
    probe <- signal (VStr "align probe")
    base <- signal (VStr "base")
    anchor <- signal (VStr "anchor")
    fit <- signal (VStr "fit")
    plain <- signal (VStr "plain probe")

    root <-
      column
        [Align AlignStretch, A11yId "root"]
        [ -- the center trio
          column
            [Align AlignCenter, A11yId "centered"]
            [ labelBound probe, -- label#0
              buttonOn "mid" (return ()),
              -- the baseline trio
              row
                [Align AlignBaseline, A11yId "baseline"]
                [ labelBound base, -- label#1
                  buttonOn "tick" (return ()),
                  imageBytes tallPng
                ]
            ],
          -- row#1: the stretch pair's host
          row
            []
            [ labelBound anchor, -- label#2
              column
                [Grow 1, Align AlignStretch, A11yId "fitcol"]
                [ labelBound fit, -- label#3
                  buttonOn "wide" (return ())
                ]
            ],
          -- row@plain: NO align, so the core's centre default is what the
          -- scene reads
          row
            [A11yId "plain"]
            [ labelBound plain [A11yId "plainlabel"], -- label#4
              imageBytes tallPng
            ],
          -- column@knobs: NO align; fill opts one child out of its
          -- default and one in
          column
            [A11yId "knobs"]
            [ textarea [Fill False, A11yId "optout"],
              button "fills" [Fill True, A11yId "fills"]
            ]
        ]
    mount root

File: guests/python/align.py
"""The align conformance scene, Python port — see
guests/rust/align.rs and tools/scenes/align.steps for the full
rationale. The stretched root spans three container children across the
window; column#1 centers children of three different natural widths;
the row aligns baselines across a label, a button, and a tall
no-baseline image whose bottom sits ON the baseline (the CSS
replaced-element rule); and row#1 hosts the grown, stretched nested
column the ruling's own construction pins.
"""

import sys

import kaya

TALL_PNG = bytes([137, 80, 78, 71, 13, 10, 26, 10, 0, 0, 0, 13, 73, 72, 68,
                  82, 0, 0, 0, 2, 0, 0, 0, 64, 8, 2, 0, 0, 0, 191,
                  68, 49, 20, 0, 0, 0, 18, 73, 68, 65, 84, 120, 156, 99, 8,
                  8, 138, 2, 34, 134, 81, 106, 104, 82, 0, 67, 50, 126, 1, 49,
                  1, 65, 124, 0, 0, 0, 0, 73, 69, 78, 68, 174, 66, 96, 130])

app = kaya.App()

with app.window():
    probe = kaya.signal("align probe")
    base = kaya.signal("base")
    anchor = kaya.signal("anchor")
    fit = kaya.signal("fit")
    plain = kaya.signal("plain probe")

    with kaya.column(align="stretch") as root:
        root.a11y_id("root")
        with kaya.column(align="center") as centered:  # the center trio
            centered.a11y_id("centered")
            kaya.label(bind=probe)  # label#0
            kaya.button("mid")
            with kaya.row(align="baseline") as baseline:  # the baseline trio
                baseline.a11y_id("baseline")
                kaya.label(bind=base)  # label#1
                kaya.button("tick")
                kaya.image(TALL_PNG)
        with kaya.row():  # row#1: the stretch pair's host
            kaya.label(bind=anchor)  # label#2
            with kaya.column(grow=1, align="stretch") as fitcol:
                fitcol.a11y_id("fitcol")
                kaya.label(bind=fit)  # label#3
                kaya.button("wide")
        # row@plain: NO align, so the core's centre default is what the
        # scene reads
        with kaya.row() as plain_row:
            plain_row.a11y_id("plain")
            kaya.label(bind=plain).a11y_id("plainlabel")  # label#4
            kaya.image(TALL_PNG)

sys.exit(app.run())
